internal/testrunner/runners/pipeline/test_result.go

```diff
@@ -70,87 +65,62 @@ func compareResults(testCasePath string, config *testConfig, result *testResult)
 		return errors.Wrap(err, "marshalling expected test results failed")
 	}
 
-	report := diffUlite(string(expected), string(actual), diffContext)
+	report, err := diffJson(expected, actual)
+	if err != nil {
+		return errors.Wrap(err, "comparing expected test result")
+	}
 	if report != "" {
 		return testrunner.ErrTestCaseFailed{
 			Reason:  "Expected results are different from actual ones",
 			Details: report,
 		}
 	}
+
 	return nil
 }
 
-// diffUlite implements a unified diff-like rendering with u lines of context.
-// It differs from a complete unified diff in that it does not provide the length
-// of the chunk, so it cannot be used to generate patches, but can be used for
-// human inspection.
-func diffUlite(a, b string, u int) string {
-	chunks := diff.DiffChunks(strings.Split(a, "\n"), strings.Split(b, "\n"))
-	if len(chunks) == 0 {
-		return ""
+func compareJsonNumbers(a, b json.Number) bool {
+	if a == b {
+		// Equal literals, so they are the same.
+		return true
 	}
-
-	buf := new(bytes.Buffer)
-	fmt.Fprintf(buf, "--- want\n+++ got\n")
-	gotLine := 1
-	wantLine := 1
-
-	for i, c := range chunks {
-		if i == 0 && (len(c.Added) != 0 || len(c.Deleted) != 0) {
-			fmt.Fprintf(buf, "@@ -%d +%d @@\n", wantLine, gotLine)
+	if inta, err := a.Int64(); err == nil {
+		if intb, err := b.Int64(); err == nil {
+			return inta == intb
 		}
-		var change bool
-		for _, line := range c.Added {
-			change = true
-			fmt.Fprintf(buf, "+ %s\n", line)
+		if floatb, err := b.Float64(); err == nil {
+			return float64(inta) == floatb
 		}
-		gotLine += len(c.Added)
-
-		for _, line := range c.Deleted {
-			change = true
-			fmt.Fprintf(buf, "- %s\n", line)
+	} else if floata, err := a.Float64(); err == nil {
+		if intb, err := b.Int64(); err == nil {
+			return floata == float64(intb)
 		}
-		wantLine += len(c.Deleted)
-
-		var used int
-		if change {
-			used = min(len(c.Equal), u)
-			for _, line := range c.Equal[:used] {
-				fmt.Fprintf(buf, "  %s\n", line)
-			}
-		}
-		if i < len(chunks)-1 {
-			next := chunks[i+1]
-			if len(next.Added) != 0 || len(next.Deleted) != 0 {
-				off := max(used, len(c.Equal)-u)
-				if off < len(c.Equal) {
-					if i == 0 || 2*u < len(c.Equal) {
-						fmt.Fprintf(buf, "@@ -%d +%d @@\n", wantLine+off, gotLine+off)
-					}
-					for _, line := range c.Equal[off:] {
-						fmt.Fprintf(buf, "  %s\n", line)
-					}
-				}
-			}
+		if floatb, err := b.Float64(); err == nil {
+			return floata == floatb
 		}
-		gotLine += len(c.Equal)
-		wantLine += len(c.Equal)
 	}
-	return strings.TrimRight(buf.String(), "\n")
+	return false
 }
 
-func min(a, b int) int {
-	if a < b {
-		return a
-	}
-	return b
-}
+func diffJson(want, got []byte) (string, error) {
+	opts := jsondiff.DefaultConsoleOptions()
+
+	// Remove colored output.
+	opts.Added = jsondiff.Tag{}
+	opts.Removed = jsondiff.Tag{}
+	opts.Changed = jsondiff.Tag{}
+	opts.Skipped = jsondiff.Tag{}
```

Comment:
This loses information about the change when it's rendered to the page. Can we have some indication in the output about the kind of diff?

Reply:
Umm, to rendering on what page?
These tags are used to add formatting, like colors, or html tags if rendered in a website, but without them it should be enough to see what is wrong. I am setting them to the zero values to avoid having colors in CI or other non-ttys, that are not going to render them properly.
Could you give an example of the problematic output and what would you expect?

Reply:
Some sigil indicating the direction of the change would be fine. I've tried to find the PR I was looking at today that made this an issue, but can't find it. Looking at the diff, I was not able to tell what was happening. An indication of whether the diffs were additions/deletions/changes would have solved that.

Reply:
Umm, I think that changes are clear, they are indicated as in the example in the description:
```
                    "sequence_number": 6920071768563516000 => 6920071768563516860,
```
But it is true that current diff doesn't distinguish between additions and deletions, and that may be confusing.

On the other hand, from the shared stacktrace it looks clear that there is some kind of problem with `host.hostname`.

I have opened a PR to add +/- tags before additions and deletions, let me know if you prefer this format: https://github.com/elastic/elastic-package/pull/770

Reply:
Thank you for doing this. While I was looking into representing the problem better by code I found an issue with the underlying package; it does not give a minimal diff. This can be seen here: https://go.dev/play/p/lTcn7sGwdnv. The upshot is that there is no attempt to map corresponding elements within a list, which is unfortunate since all the outputs that it needs to give information on are lists.

Reply:
Umm, yeah, not very nice in this case. Let's see if we find more problems with this package, we can try others if this one doesn't work well at the end.

Reply:
Thanks